library/ecdh.c

```diff
@@ -191,19 +191,28 @@ int mbedtls_ecdh_make_params( mbedtls_ecdh_context *ctx, size_t *olen,
 {
     int ret;
     size_t grp_len, pt_len;
+#if !defined(MBEDTLS_ECDH_GEN_PUBLIC_ALT)
```

Comment:
I think it would be better to change this to `#if defined(MBEDTLS_ECP_RESTARTABLE)`, and below to call `mbedtls_ecdh_gen_public` under `#if defined(MBEDTLS_ECP_RESTARTABLE)` instead of `#if defined(MBEDTLS_ECDH_GEN_PUBLIC_ALT)`. That way the data flow is easier to follow in the “base case” (no restartable, no alt implementations). It may help compilers eliminate a little RAM usage in the base case as well.

The same applies to the other functions with a similar idiom.

Reply:
Yes, I agree

Reply:
This breaks CI, because the context is used even if `MBEDTLS_ECP_RESTARTABLE` is not defined.

Reply:
Not if you change the test around the call to `mbedtls_ecdh_gen_public`. That's the only place where this function uses `rs_ctx`.

Reply:
no, the actual call `ecdh_gen_public_restartable` is used regardless whether restarbale feature is used or not.
Unless you want to call `mbedtls_ecdh_gen_public` when restartable is not used?

Reply:
> Unless you want to call mbedtls_ecdh_gen_public when restartable is not used?

Yes, that's what I requested above.

Reply:
I disagree here - I think the restartable ECC feature was implemented in a way that the `xxx_restartable()` functions would be called regardless of whether restartable operations are enabled statically or dynamically, with the restart context parameter deciding whether the feature is actually used or not. I think @RonEld's original approach of guarding the calls through `MBEDTLS_XXX_ALT` suited that better, and it didn't change the existing code in case the `MBEDTLS_XXX_ALT` feature wasn't used.

Reply:
As I mentioned in https://github.com/ARMmbed/mbedtls/pull/2165#issuecomment-436014239 I think that having a positive flow is much readable. note that in previous implementation as well, if `MBEDTLS_ECP_RESTARTABLE` was not defined, then `rs_ctx` is NULL anyway, and there was no difference between calling the _restartable function with a NULL pointer variable, or calling the non restartable functions, which in place calls the restartable function with NULL.
For example https://github.com/ARMmbed/mbedtls/pull/2165#diff-97f71deba621674b90985642f0884ebdR122